src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -1369,6 +1404,7 @@ protected MembersAndInitializers GetMembersAndInitializers()
 
             AddDeclarationDiagnostics(diagnostics);
             diagnostics.Free();
+            _lazyDeclaredMembersAndInitializers = null;
```

Comment:
>_lazyDeclaredMembersAndInitializers = null; [](start = 12, length = 43)

I think we should move this right after ```Interlocked.CompareExchange``` call above. I think it is possible that a losing thread manages to reset the field back to not null value after the winner.

Reply:
Moving it closer may help reduce the frequency for this situation, but would still have the same correctness issue :-/
I need to think more about this...

---
In reply to: [562992614](https://github.com/dotnet/roslyn/pull/50650#discussion_r562992614) [](ancestors = 562992614)

Reply:
Perhaps set `_lazyDeclaredMembersAndInitializers` to a non-null sentinel value rather than resetting it.

---
In reply to: [563001344](https://github.com/dotnet/roslyn/pull/50650#discussion_r563001344) [](ancestors = 563001344,562992614)

Reply:
@cston Yes, thanks! I'd arrived at the same idea. I'm using sentinel as starting value and null as final value. I think that works :-)

---
In reply to: [563012722](https://github.com/dotnet/roslyn/pull/50650#discussion_r563012722) [](ancestors = 563012722,563001344,562992614)